_posts/2023-09-05-wallclock-profiler.md

```diff
@@ -0,0 +1,219 @@
+
+### Tracer managed thread filter
+
+High thread counts also can also reduce sample relevance, without a thread filter. 
+For latency analysis it would be ideal to bias towards threads on the critical path. 
+There are two kinds of noncritical threads for typical applications:
+
+1. idle threads in overprovisioned thread pools, each waiting for items to appear in a queue.
+2. background threads such as metrics reporters.
+
+In the screenshot below, at least half of the samples are of the 200 workers (which is too many for this application) waiting for work to arrive on a `LinkedBlockingQueue`. 
+
+![all threads](/assets/2023/09/wallclock-profiler/all-threads-datadog-flamegraph-all-activity.png)
+
+This isn't very useful information: while fine-tuning the size of this thread pool would be generally beneficial, they aren't blocking the threads processing requests. 
+
+What's worse is when the samples are rendered as a timeline of a particular trace, the coverage is quite sparse (the grey bars are `PARKED` walltime samples, the blue bars are CPU samples). 
+
+![all timeline trace filter](/assets/2023/09/wallclock-profiler/all-threads-datadog-timeline-trace-filter.png)
+
+Sampling idle or background activity reduces the coverage of traces.
+
+Both of these problems can be solved by only sampling threads which have an active trace context set up by the tracer.
+Firstly, tracing is designed to record request latency, so if there is a trace context associated with a thread for a time period, there is a good chance the work is being done to complete a request.
+Thanks to context propagation, which propagates trace context into thread pools, any thread which ends up working to complete the request can be sampled.
+Secondly, biasing sampling towards traced activity helps _focus_ the profiler on explaining trace latency, so we have the situation where the tracer can tell you how long a request took, and there's a good chance that there will be wallclock samples to explain why it took so long.
+Whenever we have too much work to do, we can do a better job by prioritising some tasks over others, and it's the same for sampling threads.
+
+This results in a flamegraph focused on latency-critical work:
+
+![traced threads](/assets/2023/09/wallclock-profiler/thread-filter-datadog-flamegraph-all-activity.png)
+
+and when we look at the timeline of a particular trace, we have much better coverage (i.e. more grey bars for more threads)
+
+![filtered timeline trace filter](/assets/2023/09/wallclock-profiler/thread-filter-datadog-timeline-trace-filter.png)
+
+This is implemented by repurposing async-profiler's thread filter, so that it ony applied to the wallclock profiler and is controlled by the tracer. 
```

Comment:
```suggestion
This is implemented by repurposing async-profiler's thread filter, so that it is only applied to the wallclock profiler and is controlled by the tracer. 
```